segmentstore/storage/src/main/java/io/pravega/segmentstore/storage/chunklayer/ReadIndexCache.java

```diff
@@ -0,0 +1,382 @@
+     * Current generation of cache entries.
+     */
+    private final AtomicLong currentGeneration = new AtomicLong();
+
+    /**
+     * Lowest generation of cache entries.
+     */
+    private final AtomicLong oldestGeneration = new AtomicLong();
+
+    /**
+     * Total number of chunks in the cache.
+     */
+    private final AtomicInteger totalChunkCount = new AtomicInteger();
+
+    /**
+     * Index of chunks for a segment by their start offsets.
+     */
+    private final ConcurrentHashMap<String, SegmentReadIndex> segmentsToReadIndexMap = new ConcurrentHashMap<String, SegmentReadIndex>();
+
+    /**
+     * Constructor.
+     *
+     * @param maxIndexedSegments
+     * @param maxIndexedChunksPerSegment
+     * @param maxIndexedChunks
+     */
+    public ReadIndexCache(int maxIndexedSegments, int maxIndexedChunksPerSegment, int maxIndexedChunks) {
+        this.maxIndexedChunksPerSegment = maxIndexedChunksPerSegment;
+        this.maxIndexedSegments = maxIndexedSegments;
+        this.maxIndexedChunks = maxIndexedChunks;
+    }
+
+    /**
+     * Retrieves the read index for given segment.
+     *
+     * @param streamSegmentName Name of the segment.
+     * @return Read index correpsonding to the given segment. A new empty index is created if it doesn't already exist.
+     */
+    private SegmentReadIndex getSegmentReadIndex(String streamSegmentName) {
+        SegmentReadIndex readIndex;
+        if (segmentsToReadIndexMap.containsKey(streamSegmentName)) {
+            readIndex = segmentsToReadIndexMap.get(streamSegmentName);
+        } else {
+            // Evict segments if required.
+            if (maxIndexedSegments < segmentsToReadIndexMap.size() + 1 || maxIndexedChunks < totalChunkCount.get() + 1) {
+                evictSegmentsFromOldestGeneration();
+            }
+            val newReadIndex = SegmentReadIndex.builder()
+                    .chunkIndex(new ConcurrentSkipListMap<Long, SegmentReadIndexEntry>())
+                    .generation(currentGeneration.get())
+                    .build();
+            val oldReadIndex = segmentsToReadIndexMap.putIfAbsent(streamSegmentName, newReadIndex);
+            readIndex = null != oldReadIndex ? oldReadIndex : newReadIndex;
+        }
+
+        return readIndex;
+    }
+
+    /**
+     * Gets total number of chunks in cache.
+     *
+     * @return
+     */
+    public int getTotalChunksCount() {
+        return totalChunkCount.get();
+    }
+
+    /**
+     * Gets total number of segments in cache.
+     *
+     * @return
+     */
+    public int getTotalSegmentCount() {
+        return segmentsToReadIndexMap.size();
+    }
+
+    /**
+     * Gets oldest generation in cache.
+     *
+     * @return
+     */
+    public long getOldestGeneration() {
+        return oldestGeneration.get();
+    }
+
+    /**
+     * Gets current generation of cache.
+     *
+     * @return
+     */
+    public long getCurrentGeneration() {
+        return currentGeneration.get();
```

Comment:
That will help with eviction of segments but not chunks.
This data is used only for finding first chunk to read from in case of a read. (and also helps with recently written last chunks so even if the segment is evicted out of memory it can still do quick look up for recently evicted segments.

Reply:
Anyways we'll revisit this soon https://github.com/pravega/pravega/issues/4902